Simbody/src/GeneralForceSubsystem.cpp

```diff
@@ -34,12 +34,461 @@
+            }
+            break;
+
+        case NonCached:
+            if (threadIndex == NonParallelForcesIndex) {
+                // Process all non-parallel forces.
+                for (Force* force : *m_enabledNonParallelForces) {
+                    const auto& impl = force->getImpl();
+                    if (!impl.dependsOnlyOnPositions()) {
+                        impl.calcForce(*m_state,
+                                *m_rigidBodyForces, *m_particleForces,
+                                *m_mobilityForces);
+                    }
+                }
+            } else {
+                // Process a single parallel force. Subtract 1 from index b/c
+                // we use 0 for the non-parallel forces.
+                const auto& impl =
+                    m_enabledParallelForces->getElt(threadIndex-1)->getImpl();
+                    if (!impl.dependsOnlyOnPositions()) {
+                        impl.calcForce(*m_state,
+                                m_rigidBodyForcesLocal.upd(), m_particleForcesLocal.upd(),
+                                m_mobilityForcesLocal.upd());
+                    }
+            }
+            break;
+        }
+    }
+    
+    //Once a thread has finished it's force calculations, we add in the thread's
+    //contribution into the cached force arrays in the State
+    void finish() override {
+        *m_rigidBodyForces += m_rigidBodyForcesLocal.get();
+        *m_particleForces += m_particleForcesLocal.get();
+        *m_mobilityForces += m_mobilityForcesLocal.get();
+
+        if (m_mode == CachedAndNonCached) {
+            *m_rigidBodyForceCache += m_rigidBodyForceCacheLocal.get();
+            *m_particleForceCache += m_particleForceCacheLocal.get();
+            *m_mobilityForceCache += m_mobilityForceCacheLocal.get();
+        }
+    }
+private:
+    Mode m_mode;
+
+    ReferencePtr<const State> m_state;
+
+    // Constant state-cache variables for the enabled parallel and non-parallel
+    // forces.
+    ReferencePtr<const Array_<Force*>> m_enabledNonParallelForces;
+    ReferencePtr<const Array_<Force*>> m_enabledParallelForces;
+
+    // ReferencePtrs that point to caches in the state; We eventually add our
+    // thread-local result to these vectors.
+    ReferencePtr<Vector_<SpatialVec>> m_rigidBodyForces;
+    ReferencePtr<Vector_<Vec3>> m_particleForces;
+    ReferencePtr<Vector> m_mobilityForces;
+
+    ReferencePtr<Vector_<SpatialVec>> m_rigidBodyForceCache;
+    ReferencePtr<Vector_<Vec3>> m_particleForceCache;
+    ReferencePtr<Vector> m_mobilityForceCache;
+    
+    // These variables are local to a thread. They are set to their default
+    // value when the threads are spawned. We use them to keep track of each
+    // thread's contribution that we will later add in to the final state cache.
+    ThreadLocal<Vector_<SpatialVec>> m_rigidBodyForcesLocal;
+    ThreadLocal<Vector_<Vec3>> m_particleForcesLocal;
+    ThreadLocal<Vector> m_mobilityForcesLocal;
```

Comment:
When I try to do so I get weird linker errors I can't see how to fix online.

>/usr/bin/ld: CMakeFiles/SimTKsimbody.dir/__/src/GeneralForceSubsystem.cpp.o: relocation R_X86_64_PC32 against undefined symbol `_ZN12_GLOBAL__N_122CalcForcesParallelTask22m_rigidBodyForcesLocalE' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value